src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala

```diff
@@ -112,31 +115,28 @@ trait RegularSync {
 
   private def handleBlockBodies(peer: ActorRef, m: Seq[BlockBody]) = {
     if (m.nonEmpty) {
-      val result = headersQueue.zip(m).map { case (h, b) => blockValidator(h, b) }
+      val result = headersQueue.zip(m).map { case (h, b) => validators.blockValidator.validateHeaderAndBody(h, b) }
 
       if (!result.exists(_.isLeft) && result.nonEmpty) {
         val blocks = result.collect { case Right(b) => b }
 
         blockchain.getBlockHeaderByHash(blocks.head.header.parentHash)
           .flatMap(b => blockchain.getTotalDifficultyByHash(b.hash)) match {
-          case Some(td) =>
-            var currentTd = td
-            val newBlocks = blocks.map { b =>
-              val blockHashToDelete = blockchain.getBlockHeaderByNumber(b.header.number).map(_.hash).filter(_ != b.header.hash)
-              blockchain.save(b)
-              appStateStorage.putBestBlockNumber(b.header.number)
-              currentTd += b.header.difficulty
-              blockchain.save(b.header.hash, currentTd)
-              blockHashToDelete.foreach(blockchain.removeBlock)
-
-              NewBlock(b, currentTd)
+          case Some(blockParentTd) =>
+            val (newBlocks, error) = processBlocks(blocks, blockParentTd)
+
+            if(newBlocks.nonEmpty){
+              context.self ! BroadcastBlocks(newBlocks)
+              log.info(s"got new blocks up till block: ${newBlocks.last.block.header.number} " +
+                s"with hash ${Hex.toHexString(newBlocks.last.block.header.hash.toArray[Byte])}")
             }
 
-            context.self ! BroadcastBlocks(newBlocks)
-            log.info(s"got new blocks up till block: ${newBlocks.last.block.header.number} " +
-              s"with hash ${Hex.toHexString(newBlocks.last.block.header.hash.toArray[Byte])}")
+            if(error.isDefined){
+              val numberBlockFailed = blocks.head.header.number + newBlocks.length
+              resumeWithDifferentPeer(peer, reason = s"a block execution error: ${error.get.toString}, in block $numberBlockFailed")
+            }
           case None =>
-            log.error("no total difficulty for latest block")
+            log.error("no total difficulty for latest block") //FIXME: How do we handle this error on our blockchain?
```

Comment:
I agree :+1: I think it should only be caused by data corruption (or by a bug), I'll create a task to handle this error